File: medium/3.longest-substring-without-repeating-characters.java
import java.util.HashSet;

/*
 * @lc app=leetcode id=3 lang=java
 *
 * [3] Longest Substring Without Repeating Characters
 */

// @lc code=start
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> set = new HashSet<>();
        char[] chars = s.toCharArray();

        int l=0,r=0;
        int ans = 0;
        while (r<chars.length) {
            if (!set.contains(chars[r])) {
                set.add(chars[r]);
                ans = Math.max(ans, r-l+1);
                r++;
            } else {
                // c is a duplicated char, move the left side
                while (set.contains(chars[r])) {
                    set.remove(chars[l]);
                    l++;
                }
            }
        }

        return ans;

    }
}
// @lc code=end

File: medium/3.longest-substring-without-repeating-characters.py
#
# @lc app=leetcode id=3 lang=python3
#
# [3] Longest Substring Without Repeating Characters
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        l = 0
        ans = 0
        charsInWindow = set()
        for r in range(len(s)):
            while s[r] in charsInWindow:
                charsInWindow.remove(s[l])
                l += 1
                
            charsInWindow.add(s[r])
            ans = max(ans, r-l+1)

        return ans
        
# @lc code=end

